runtime/vm/createramclass.cpp

```diff
@@ -1622,30 +1637,48 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
 static VMINLINE BOOLEAN
-loadFlattenableFieldValueClasses(J9VMThread *vmThread, J9ClassLoader *classLoader, J9ROMClass *romClass, UDATA classPreloadFlags)
+loadFlattenableFieldValueClasses(J9VMThread *vmThread, J9ClassLoader *classLoader, J9ROMClass *romClass, UDATA classPreloadFlags, J9Class **valueTypesOut)
 {
-	struct J9ROMFieldWalkState fieldWalkState;
-	memset(&fieldWalkState, 0, sizeof(fieldWalkState));
-
+	J9ROMFieldWalkState fieldWalkState = {0};
+	J9SRP *valueTypeClasses = J9ROMCLASS_VALUETYPECLASSES(romClass);
+	UDATA valueTypeClassCount = romClass->valueTypeClassCount;
+	UDATA index = 0;
 	/* iterate over fields and load classes of fields marked flattenable */
-	J9ROMFieldShape *field = romFieldsStartDo(romClass, &fieldWalkState);
-	while (NULL != field) {
+	for (J9ROMFieldShape *field = romFieldsStartDo(romClass, &fieldWalkState); NULL != field; field = romFieldsNextDo(&fieldWalkState)) {
 		const U_32 modifiers = field->modifiers;
 		J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(field);
-
-		if (modifiers & J9AccFlattenable) {
-			if ('[' == J9UTF8_DATA(signature)[0]) {
-				// TODO: array case
-				// fieldClass = internalFindClassUTF8(vmThread, J9UTF8_DATA(signature), J9UTF8_LENGTH(signature), classLoader, classPreloadFlags);
+		if (J9_ARE_ALL_BITS_SET(modifiers, J9AccFlattenable) && ('[' != J9UTF8_DATA(signature)[0])) {
+			Assert_VM_true('L' == J9UTF8_DATA(signature)[0]);
+			if (J9_ARE_ALL_BITS_SET(modifiers, J9AccStatic)) {
+				/* The static fields will be loaded later during class initialization */
+				valueTypesOut[index] = (J9Class *) signature;
 			} else {
-				Assert_VM_true('L' == J9UTF8_DATA(signature)[0]);
-				/* skip fieldSignature's L and ; to have only CLASSNAME required for internalFindClassUTF8 */
-				internalFindClassUTF8(vmThread, &J9UTF8_DATA(signature)[1], J9UTF8_LENGTH(signature)-2, classLoader, classPreloadFlags);
+				J9Class *currentClass = internalFindClassUTF8(vmThread, &J9UTF8_DATA(signature)[1], J9UTF8_LENGTH(signature)-2, classLoader, classPreloadFlags);
+				if (J9_ARE_NO_BITS_SET(currentClass->romClass->modifiers, J9AccValueType)) {
+					setCurrentExceptionForBadClass(vmThread, J9ROMCLASS_CLASSNAME(romClass), J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR);
+					return FALSE;
+				}
+				J9UTF8 *valueTypeClassName = NULL;
+				BOOLEAN found = FALSE;
+				for (UDATA i = 0; i < valueTypeClassCount; i++) {
+					valueTypeClassName = NNSRP_GET(valueTypeClasses[i], J9UTF8*);
+					if (J9UTF8_EQUALS(signature, valueTypeClassName)) {
+						valueTypesOut[index] = (J9Class *) (((uintptr_t) currentClass) | VALUE_TYPE_INSTANCE_FIELD_TAG);
+						found = TRUE;
+						break;
+					}
+				}
+				if (!found) {
+					setCurrentExceptionForBadClass(vmThread, J9ROMCLASS_CLASSNAME(romClass), J9VMCONSTANTPOOL_JAVALANGCLASSFORMATERROR);
+					return FALSE;
+				}
 			}
+			index++;
 		}
-		field = romFieldsNextDo(&fieldWalkState);
 	}
-
+	if (index < valueTypeClassCount) {
+		valueTypesOut[index] = NULL;
```

Comment:
Is it ok to remove this since the valueTypesOut array is initialized to 0 before it is passed to this function? Either with = {0} or calloc

Reply:
yes